src/Skia/Avalonia.Skia/DrawingContextImpl.cs

```diff
@@ -79,363 +138,554 @@ public void DrawGeometry(IBrush brush, Pen pen, IGeometryImpl geometry)
             }
         }
 
-        internal struct PaintWrapper : IDisposable
+        /// <inheritdoc />
+        public IRenderTargetBitmapImpl CreateLayer(Size size)
         {
-            //We are saving memory allocations there
-            //TODO: add more disposable fields if needed
-            public readonly SKPaint Paint;
+            var normalizedDpi = new Vector(_dpi.X / SkiaPlatform.DefaultDpi.X, _dpi.Y / SkiaPlatform.DefaultDpi.Y);
+            var pixelSize = size * normalizedDpi;
 
-            private IDisposable _disposable1;
-            private IDisposable _disposable2;
+            return CreateRenderTarget((int) pixelSize.Width, (int) pixelSize.Height, _dpi);
+        }
 
-            public IDisposable ApplyTo(SKPaint paint)
-            {
-                var state = new PaintState(paint, paint.Color, paint.Shader);
+        /// <inheritdoc />
+        public void PushClip(Rect clip)
+        {
+            Canvas.Save();
+            Canvas.ClipRect(clip.ToSKRect());
+        }
 
-                paint.Color = Paint.Color;
-                paint.Shader = Paint.Shader;
+        /// <inheritdoc />
+        public void PopClip()
+        {
+            Canvas.Restore();
+        }
 
-                return state;
-            }
+        /// <inheritdoc />
+        public void PushOpacity(double opacity)
+        {
+            _opacityStack.Push(_currentOpacity);
+            _currentOpacity *= opacity;
+        }
 
-            public void AddDisposable(IDisposable disposable)
-            {
-                if (_disposable1 == null)
-                    _disposable1 = disposable;
-                else if (_disposable2 == null)
-                    _disposable2 = disposable;
-                else
-                    throw new InvalidOperationException();
-            }
+        /// <inheritdoc />
+        public void PopOpacity()
+        {
+            _currentOpacity = _opacityStack.Pop();
+        }
 
-            public PaintWrapper(SKPaint paint)
+        /// <inheritdoc />
+        public virtual void Dispose()
+        {
+            if (_disposables == null)
             {
-                Paint = paint;
-                _disposable1 = null;
-                _disposable2 = null;
+                return;
             }
 
-            public void Dispose()
+            foreach (var disposable in _disposables)
             {
-                Paint?.Dispose();
-                _disposable1?.Dispose();
-                _disposable2?.Dispose();
+                disposable?.Dispose();
             }
         }
 
-        internal PaintWrapper CreatePaint(IBrush brush, Size targetSize)
+        /// <inheritdoc />
+        public void PushGeometryClip(IGeometryImpl clip)
         {
-            SKPaint paint = new SKPaint();
-            var rv = new PaintWrapper(paint);
-            paint.IsStroke = false;
+            Canvas.Save();
+            Canvas.ClipPath(((StreamGeometryImpl) clip).EffectivePath);
```

Comment:
IGeometryImpl in Skia has two implementations StreamGeometryImpl and TransformedGeometryImpl (and one for MockStreamGeometryImpl). I'm a bit scared unexpected NullReferenceException here.

Reply:
I can change this to an `is` check to avoid the potential runtime crash

Reply:
And may be use base class GeometryImpl (from Avalonia.Skia)?
I sure, I have much less experience and know project worse than most of you, but it seems to me, what there may be problems in the future with cast like this.